File: src/instruction.rs
use borsh::{BorshDeserialize, BorshSerialize};
use solana_program::{
    instruction::{AccountMeta, Instruction},
    program_error::ProgramError,
    pubkey::Pubkey,
    system_program,
};

#[derive(Debug, PartialEq)]
pub enum TradeYardInstruction {
    Sell,
    Buy,
    Cancel,
}

impl TradeYardInstruction {
    pub fn unpack(instruction_data: &[u8]) -> Result<(Self, Args), ProgramError> {
        let payload = Payload::try_from_slice(instruction_data)?;

        let instruction = match payload.instruction {
            0 => Self::Sell,
            1 => Self::Buy,
            2 => Self::Cancel,
            _ => return Err(ProgramError::InvalidInstructionData),
        };

        Ok((instruction, payload.args))
    }
}

#[derive(BorshDeserialize, BorshSerialize, Debug)]
pub struct Payload {
    pub instruction: u8,
    pub args: Args,
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Clone, Copy, Debug)]
pub struct Args {
    pub lamports: Option<u64>,
    pub metadata_bump: Option<u8>,
}

pub fn sell(
    seller: &Pubkey,
    program_item_wallet: &Pubkey,
    mint: &Pubkey,
    payment: &Pubkey,
    lamports: u64,
) -> Instruction {
    let (item_metadata_addr, item_metadata_bump) = crate::find_item_metadata_address(mint);

    Instruction::new_with_borsh(
        crate::id(),
        &Payload {
            instruction: 0, // Sell
            args: Args {
                lamports: Some(lamports),
                metadata_bump: Some(item_metadata_bump),
            },
        },
        vec![
            AccountMeta::new(*seller, true),
            AccountMeta::new_readonly(*program_item_wallet, false),
            AccountMeta::new_readonly(*mint, false),
            AccountMeta::new(item_metadata_addr, false),
            AccountMeta::new_readonly(*payment, false),
            AccountMeta::new_readonly(system_program::id(), false),
        ],
    )
}

pub fn cancel(
    seller: &Pubkey,
    program_wallet: &Pubkey,
    seller_wallet: &Pubkey,
    mint: &Pubkey
) -> Instruction {
    Instruction::new_with_borsh(
        crate::id(),
        &Payload {
            instruction: 2, // Sell
            args: Args {
                lamports: None,
                metadata_bump: None,
            },
        },
        vec![
            AccountMeta::new_readonly(*seller, true),
            AccountMeta::new(crate::find_item_metadata_address(mint).0, false),
            AccountMeta::new_readonly(spl_token::id(), false),
            AccountMeta::new(*program_wallet, false),
            AccountMeta::new(*seller_wallet, false),
            AccountMeta::new_readonly(crate::find_item_address(mint).0, false),
        ],
    )
}

pub fn buy(
    buyer: &Pubkey,
    buyer_wallet: &Pubkey,
    buyer_receiver_wallet: &Pubkey,
    program_wallet: &Pubkey,
    payment_wallet: &Pubkey,
    metadata: &Pubkey,
    item: &Pubkey,
    ) -> Instruction {
    Instruction::new_with_borsh(
        crate::id(),
        &Payload {
            instruction: 1, // Sell
            args: Args {
                lamports: None,
                metadata_bump: None,
            },
        },
        vec![
            AccountMeta::new(*buyer, true),
            AccountMeta::new(*buyer_wallet, false),
            AccountMeta::new(*buyer_receiver_wallet, false),
            AccountMeta::new(*program_wallet, false),
            AccountMeta::new(*payment_wallet, false),
            AccountMeta::new(*metadata, false),
            AccountMeta::new_readonly(spl_token::id(), false),
            AccountMeta::new_readonly(*item, false),
        ],
    )
}

File: js/src/instruction.ts
import {
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";
import { PROGRAM_ID } from "./program";
import { Args, Payload, schema } from "./schema";
import { getAssociatedTokenAddress, TOKEN_PROGRAM_ID } from "@solana/spl-token";
import { serialize } from "borsh";
import { findItemAddress, findItemMetadataAddress } from "./utils";
import { Buffer } from "buffer"

/**
 * Create sell instruction
 */
async function createSellInstruction(
  seller: PublicKey,
  mint: PublicKey,
  sellerPaymentWallet: PublicKey,
  lamports: number
) {
  const [itemAddr] = await findItemAddress(mint);
  const [itemMetadataAddr, itemMetadataBump] = await findItemMetadataAddress(
    mint
  );

  const payload = new Payload({
    instruction: 0,
    args: new Args({
      lamports,
      metadata_bump: itemMetadataBump,
    }),
  });

  const programItemWallet = await getAssociatedTokenAddress(
    mint,
    itemAddr,
    true
  );

  const data = Buffer.from(serialize(schema, payload));

  const instruction = new TransactionInstruction({
    programId: PROGRAM_ID,
    data,
    keys: [
      {
        pubkey: seller,
        isSigner: true,
        isWritable: false,
      },
      {
        pubkey: programItemWallet,
        isSigner: false,
        isWritable: false,
      },
      {
        pubkey: mint,
        isSigner: false,
        isWritable: false,
      },
      {
        pubkey: itemMetadataAddr,
        isSigner: false,
        isWritable: true,
      },
      {
        pubkey: sellerPaymentWallet,
        isSigner: false,
        isWritable: false,
      },
      {
        pubkey: SystemProgram.programId,
        isSigner: false,
        isWritable: false,
      },
    ],
  });

  return instruction;
}

/**
 * Create cancel instruction
 */
async function createCancelInstruction(
  seller: PublicKey,
  mint: PublicKey,
  sellerItemWallet: PublicKey
) {
  const [itemAddr] = await findItemAddress(mint);
  const [itemMetadataAddr] = await findItemMetadataAddress(mint);

  const payload = new Payload({
    instruction: 2,
    args: new Args({
      lamports: null,
      metadata_bump: null,
    }),
  });

  const programItemWallet = await getAssociatedTokenAddress(
    mint,
    itemAddr,
    true
  );

  const data = Buffer.from(serialize(schema, payload));

  const instruction = new TransactionInstruction({
    programId: PROGRAM_ID,
    data,
    keys: [
      {
        pubkey: seller,
        isSigner: true,
        isWritable: false,
      },
      {
        pubkey: itemMetadataAddr,
        isSigner: false,
        isWritable: true,
      },
      {
        pubkey: TOKEN_PROGRAM_ID,
        isSigner: false,
        isWritable: false,
      },
      {
        pubkey: programItemWallet,
        isSigner: false,
        isWritable: true,
      },
      {
        pubkey: sellerItemWallet,
        isSigner: false,
        isWritable: true,
      },
      {
        pubkey: itemAddr,
        isSigner: false,
        isWritable: false,
      },
    ],
  });

  return instruction;
}

/**
 * Create buy instruction
 */
async function createBuyInstruction(
  buyer: PublicKey,
  buyerPaymentWallet: PublicKey,
  buyerItemWalelt: PublicKey,
  sellerPaymentWallet: PublicKey,
  mint: PublicKey
) {
  const [itemAddr] = await findItemAddress(mint);
  const [itemMetadataAddr] = await findItemMetadataAddress(mint);

  const payload = new Payload({
    instruction: 1,
    args: new Args({
      lamports: null,
      metadata_bump: null,
    }),
  });

  const programItemWallet = await getAssociatedTokenAddress(
    mint,
    itemAddr,
    true
  );

  const data = Buffer.from(serialize(schema, payload));

  const instruction = new TransactionInstruction({
    programId: PROGRAM_ID,
    data,
    keys: [
      {
        pubkey: buyer,
        isSigner: true,
        isWritable: false,
      },
      {
        pubkey: buyerPaymentWallet,
        isSigner: false,
        isWritable: true,
      },
      {
        pubkey: buyerItemWalelt,
        isSigner: false,
        isWritable: true,
      },
      {
        pubkey: programItemWallet,
        isSigner: false,
        isWritable: true,
      },
      {
        pubkey: sellerPaymentWallet,
        isSigner: false,
        isWritable: true,
      },
      {
        pubkey: itemMetadataAddr,
        isSigner: false,
        isWritable: true,
      },
      {
        pubkey: TOKEN_PROGRAM_ID,
        isSigner: false,
        isWritable: false,
      },
      {
        pubkey: itemAddr,
        isSigner: false,
        isWritable: false,
      },
    ],
  });

  return instruction;
}

export { createSellInstruction, createCancelInstruction, createBuyInstruction };
